File: app/src/main/java/info/staticfree/supergenpass/view/VisualHashEditText.java
package info.staticfree.supergenpass.view;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.security.NoSuchAlgorithmException;

import info.staticfree.supergenpass.R;
import info.staticfree.supergenpass.drawable.VisualHash;

public class VisualHashEditText extends EditText {
    private boolean mShowVisualHash;
    @Nullable
    private final VisualHash mVh;

    public VisualHashEditText(@NonNull Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);

        try {
            mVh = new VisualHash();
        } catch (@NonNull NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }

        init(context, attrs);
    }

    public VisualHashEditText(@NonNull Context context, AttributeSet attrs) {
        super(context, attrs);

        try {
            mVh = new VisualHash();
        } catch (@NonNull NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }

        init(context, attrs);
    }

    public VisualHashEditText(@NonNull Context context) {
        super(context);

        try {
            mVh = new VisualHash();
        } catch (@NonNull NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }

        init(context, null);
    }

    private void init(@NonNull Context context, AttributeSet attrs) {
        int h, w;

        if (isInEditMode()) {
            h = 45;
            w = 45;
        } else {
            TypedArray ta =
                    context.obtainStyledAttributes(attrs, R.styleable.VisualHashEditText);

            mShowVisualHash = ta.getBoolean(R.styleable.VisualHashEditText_showVisualHash, true);

            h = ta.getDimensionPixelSize(R.styleable.VisualHashEditText_visualHashHeight, 45);
            w = ta.getDimensionPixelSize(R.styleable.VisualHashEditText_visualHashWidth, 45);

            ta.recycle();
        }

        // this number is based on what looks good with standard edit boxes.
        if (mVh != null) {
            mVh.setBounds(0, 0, h, w);
        }

        refreshVisualHash();
    }

    public void setShowVisualHash(boolean showVisualHash) {
        mShowVisualHash = showVisualHash;
        refreshVisualHash();
    }

    private void refreshVisualHash() {
        if (mShowVisualHash) {
            setCompoundDrawables(null, null, mVh, null);
        } else {
            setCompoundDrawables(null, null, null, null);
        }
    }

    @Override
    protected void onTextChanged(@Nullable CharSequence text, int start, int lengthBefore,
            int lengthAfter) {
        super.onTextChanged(text, start, lengthBefore, lengthAfter);

        if (text != null && mVh != null) {
            mVh.setData(text.toString().getBytes());
        }
    }

    @Override
    public void setError(@Nullable CharSequence error, Drawable icon) {
        super.setError(error, icon);

        if (error == null) {
            refreshVisualHash();
        }
    }
}

File: app/src/main/java/info/staticfree/supergenpass/view/VisualHashEditText.kt
package info.staticfree.supergenpass.view

import android.content.Context
import info.staticfree.supergenpass.drawable.VisualHash
import info.staticfree.supergenpass.R
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatEditText

class VisualHashEditText : AppCompatEditText {
    private var showVisualHash = false
    private var visualHash: VisualHash? = null

    constructor(context: Context, attrs: AttributeSet?, defStyle: Int) : super(
        context,
        attrs,
        defStyle
    ) {
        init(context, attrs)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init(context, attrs)
    }

    constructor(context: Context) : super(context) {
        init(context, null)
    }

    private fun init(context: Context, attrs: AttributeSet?) {
        val h: Int
        val w: Int
        if (isInEditMode) {
            h = 45
            w = 45
        } else {
            context.obtainStyledAttributes(attrs, R.styleable.VisualHashEditText).let {
                showVisualHash = it.getBoolean(R.styleable.VisualHashEditText_showVisualHash, true)
                h = it.getDimensionPixelSize(R.styleable.VisualHashEditText_visualHashHeight, 45)
                w = it.getDimensionPixelSize(R.styleable.VisualHashEditText_visualHashWidth, 45)
                it.recycle()
            }
        }
        visualHash = VisualHash()
        visualHash?.setBounds(0, 0, h, w)

        refreshVisualHash()
    }

    fun setShowVisualHash(showVisualHash: Boolean) {
        this.showVisualHash = showVisualHash
        refreshVisualHash()
    }

    private fun refreshVisualHash() {
        if (showVisualHash) {
            setCompoundDrawables(null, null, visualHash, null)
        } else {
            setCompoundDrawables(null, null, null, null)
        }
    }

    override fun onTextChanged(
        text: CharSequence?,
        start: Int,
        lengthBefore: Int,
        lengthAfter: Int
    ) {
        super.onTextChanged(text, start, lengthBefore, lengthAfter)
        text?.let {
            visualHash?.setData(it.toString().toByteArray())
        }
    }

    override fun setError(error: CharSequence?, icon: Drawable) {
        super.setError(error, icon)

        if (error == null) {
            refreshVisualHash()
        }
    }
}